### PR TITLE
Add more opcodes tests. Add Script Debugger

### PR DESCRIPTION
**Pull request by dfoderick**
* Fixes a few edge cases in op code implementation
* Adds more test cases. It should cover everything that bitcoind tests for
* Adds a Script Debugger

The Script Debugger is a callback function that passes the contents of the stack to the caller on every step of execution of the script interpreter. This will allow users to debug script and inspect the stack as it executes. I need to finish off a few details but it will work like this:

```
      var si = Interpreter()
      si.callbackDebug = function (step, stack) {
          prettyprint(stack)       
      }
      si.verify(Script('OP_1 OP_2 OP_ADD'), Script('OP_3 OP_EQUAL'))
```
where prettyprint is just some method that prints the stack in a nice way. Example output:
```
Step 0: OP_1: 01
Step 1: OP_2: 01
Step 1: OP_2: 02
Step 2: OP_ADD: 03
Step 0: OP_3: 03
Step 0: OP_3: 03
Step 1: OP_EQUAL: 01
```

### CODE REVIEW

**Comment by ryanxcharles** [comment]
excellent, script debugger will make the "In Browser Script Debugger, Stepper and IDE" possible

...what were the previous LSHIFT tests? they were from a different codebase with a different implementation of LSHIFT?

**Comment by dfoderick** [comment]
Tests for opcodes all come from https://github.com/bitcoin-sv/bitcoin-sv/blob/d9b12a23dbf0d2afc5f488fa077d762b302ba873/src/test/opcode_tests.cpp

In the first round of testing I implemented what I could using script_tests.json but if you look at the bitcoind tests they are kind of elaborate. They test negative numbers by performing negation on the test inputs for example. So to get 100% same test coverage I had to write code. But I think it was worth it in the end. I got very close to duplicating all the bitcoind tests.

**Comment by ryanxcharles** [comment]
i believe your script debugger could be made simpler. first of all, there are two stacks (the second stack necessary for turing-completeness :) ). but rather than return a stack, you could return ... nothing. because the script interpreter object itself has the stacks already, which you can access in the callback.

```
      var si = Interpreter()
      si.callbackDebug = function () {
          prettyprint(si.pc, si.stack, si.altstack)       
      }
      si.verify(Script('OP_1 OP_2 OP_ADD'), Script('OP_3 OP_EQUAL'))
```

another option is to return the script interpreter itself into the callback:

```
      var si = Interpreter()
      si.callbackDebug = function (si) {
          prettyprint(si.pc, si.stack, si.altstack)       
      }
      si.verify(Script('OP_1 OP_2 OP_ADD'), Script('OP_3 OP_EQUAL'))
```

**Comment by ryanxcharles** [comment]
sorry, i see your "stack" object actually contains both stacks.

still, i think we should consider passing back the entire interpreter, or nothing at all (and using the interpeter object you already have), since it contains other interesting information like the possible errstr and vfExec.

**Comment by dfoderick** [comment]
> excellent, script debugger will make the "In Browser Script Debugger, Stepper and IDE" possible
> 
Yes. I have some ideas I want to see in a script debugger and this will make it possible. I don't think any of the current ones show the alt stack so I added it to this debugger.

There is still code in there for Monolithic and Magnetic forks. A low priority task would be to remove that code.

The last 2 disabled opcodes are OP_2MUL and OP_2DIV
https://reviews.bitcoinabc.org/D1608

**Comment by dfoderick** [comment]
> sorry, i see your "stack" object actually contains both stacks.
> 
> still, i think we should consider passing back the entire interpreter, or nothing at all (and using the interpeter object you already have), since it contains other interesting information like the possible errstr and vfExec.

In an earlier implementation i had interpreter as an argument but i thought caller could keep a reference to it if they need it. I am open to that change.

Oh, that makes me think one more thing. It probably should pass a copy of the stack to the callback.
I worry about someone modifying the stack object or programcounter inside the callback. They really should not be able to do that. They could mess up execution. We need to make it safe too.

**Comment by ryanxcharles** [comment]
that's a good idea. maybe it could pass in a COPY of the stack and altstack back to the callback. if the user of the callback wants to access they entire interpreter, they can just keep that themselves and then they can access errstr and vfExec.

**Comment by ryanxcharles** [comment]
one more thing - it looks a little odd to me that you have a stack object that says "main" and "alstack". maybe just pass back a "stack" and "altstack" which are copies of the original, that way they have the same variable names as inside the interpreter itself?

**Comment by dfoderick** [comment]
> one more thing - it looks a little odd to me that you have a stack object that says "main" and "altstack". maybe just pass back a "stack" and "altstack" which are copies of the original, that way they have the same variable names as inside the interpreter itself?

Ok. I will make that change.
At a minimum, the program counter should be passed to the callback, otherwise there will be endless confusion. The reason is that the Step interpreter method increments the programcounter after it runs. So after running Step, interpreter.pc points not to the instruction that was just ran but to the next instruction to be run. In other words at the time of the callback the programcounter and stack are kind of out of sync.

Bottom line, to make it easy on user and error free I will make it these parameters:
1) step.programcounter = integer, programcounter at the time of running previous step that made the stack
step.opcode = Opcode instruction, to make it easy for user to display
2) stack = copy of main stack
3) altstack = copy of alt stack

**Comment by ryanxcharles** [comment]
ok, all those choices make sense.

File: lib/script/interpreter.js [hunk 1]
```diff
@@ -203,7 +203,7 @@ Interpreter.prototype.set = function (obj) {
   this.nin = typeof obj.nin !== 'undefined' ? obj.nin : this.nin
   this.satoshisBN = obj.satoshisBN || this.satoshisBN
   this.stack = obj.stack || this.stack
-  this.altstack = obj.altack || this.altstack
```
Comment (ryanxcharles) [inline lib/script/interpreter.js]: ha, wow, good catch!
